src/util/ringdelaybuffer.cpp

```diff
@@ -0,0 +1,145 @@
+        : m_firstInputBuffer(true),
+          m_writePos(0),
+          m_buffer(bufferSize) {
+    // Set the ring delay buffer items to 0.
+    m_buffer.fill(0);
+}
+
+void RingDelayBuffer::copy(const ReadableSlice pSourceBuffer,
+        SINT sourcePos,
+        const WritableSlice pDestBuffer,
+        SINT destPos,
+        const SINT numItems) {
+    // Check to see if the copy is not contiguous.
+    if (sourcePos + numItems > pSourceBuffer.length() ||
+            destPos + numItems > pDestBuffer.length()) {
+        // Copy is not contiguous.
+        SINT firstDataBlockSize = math_min(pSourceBuffer.length() - sourcePos,
+                pDestBuffer.length() - destPos);
+
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                firstDataBlockSize);
+
+        sourcePos = (sourcePos + firstDataBlockSize) % pSourceBuffer.length();
+        destPos = (destPos + firstDataBlockSize) % pDestBuffer.length();
+
+        // The second data part is the start of the ring buffer.
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                numItems - firstDataBlockSize);
+    } else {
+        // Copy is contiguous.
+        SampleUtil::copy(pDestBuffer.data(destPos),
+                pSourceBuffer.data(sourcePos),
+                numItems);
+    }
+}
+
+SINT RingDelayBuffer::read(CSAMPLE* pBuffer, const SINT itemsToRead, const SINT delayItems) {
+    const SINT shift = itemsToRead + delayItems;
+
+    if (shift > m_buffer.size()) {
+        return 0;
+    }
+
+    SINT readPos = m_writePos - shift;
+
+    if (readPos < 0) {
+        readPos = readPos + m_buffer.size();
+    }
+
+    copy(ReadableSlice(m_buffer.data(), m_buffer.size()),
+            readPos,
+            WritableSlice(pBuffer, itemsToRead),
+            0,
+            itemsToRead);
+
+    return itemsToRead;
+}
+
+SINT RingDelayBuffer::write(const CSAMPLE* pBuffer, const SINT itemsToWrite) {
+    if (itemsToWrite > m_buffer.size()) {
+        return 0;
+    }
+
+    if (m_firstInputBuffer) {
+        // If the first input buffer is written, the first sample is on the index 0.
+        // Based on the checking of an available number of samples, the situation,
+        // that the writing will be non-contiguous cannot occur.
+        // The itemsToWrite value is multiply by 2 to
+        SampleUtil::copyWithRampingGain(m_buffer.data(), pBuffer, 0.0f, 1.0f, itemsToWrite);
+        m_firstInputBuffer = false;
+    } else {
+        copy(ReadableSlice(pBuffer, itemsToWrite),
+                0,
+                WritableSlice(m_buffer.data(), m_buffer.size()),
+                m_writePos,
+                itemsToWrite);
+    }
+
+    // Calculate the new write position. If the new write position
+    // is after the ring delay buffer end, move it around from the start
+    // of the ring delay buffer.
+    m_writePos = (m_writePos + itemsToWrite);
```

Comment:
This makes assumption of the behavior of copy() and is not correct in the case of an early return.
One option is to pass m_writePos by pointer. The same goes for itemsToRead above.

Reply:
Right, thank you for this hint. Based on the latest changes for the `copyRing()` the fix for this bug should be included.